src/hostapi/pulseaudio/pa_linux_pulseaudio_block.c

```diff
@@ -0,0 +1,186 @@
+#include <unistd.h>
+
+/*
+    As separate stream interfaces are used for blocking and callback
+    streams, the following functions can be guaranteed to only be called
+    for blocking streams.
+*/
+
+PaError PaPulseAudio_ReadStreamBlock( PaStream * s,
+                                      void *buffer,
+                                      unsigned long frames )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) s;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi =
+        l_ptrStream->hostapi;
+    PaError l_iRet = 0;
+    size_t l_lReadable = 0;
+    uint8_t *l_ptrData = (uint8_t *) buffer;
+    long l_lLength = (frames * l_ptrStream->inputFrameSize);
+
+    pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+    while( l_lLength > 0)
+    {
+        long l_read = PaUtil_ReadRingBuffer( &l_ptrStream->inputRing, l_ptrData,
+                                             l_lLength );
+        l_ptrData += l_read;
+        l_lLength -= l_read;
+        if( l_lLength > 0 )
+            pa_threaded_mainloop_wait( l_ptrStream->mainloop );
+    }
+    pa_threaded_mainloop_unlock( l_ptrStream->mainloop );
+    return paNoError;
+}
+
+
+PaError PaPulseAudio_WriteStreamBlock( PaStream * s,
+                                       const void *buffer,
+                                       unsigned long frames )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) s;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi =
+        l_ptrStream->hostapi;
+    int l_iRet = 0;
+    size_t l_lWritable = 0;
+    uint8_t *l_ptrData = (uint8_t *) buffer;
+    long l_lLength = (frames * l_ptrStream->outputFrameSize);
+    pa_operation *l_ptrOperation = NULL;
+
+    PaUtil_BeginCpuLoadMeasurement( &l_ptrStream->cpuLoadMeasurer );
+
+    while( l_lLength > 0)
+    {
+        pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+        l_lWritable = pa_stream_writable_size( l_ptrStream->outStream );
+        pa_threaded_mainloop_unlock( l_ptrStream->mainloop );
+
+        if( l_lWritable > 0 )
+        {
+           if( l_lLength < l_lWritable )
+           {
+                l_lWritable = l_lLength;
+           }
+           pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+           l_iRet = pa_stream_write( l_ptrStream->outStream,
+                                     l_ptrData,
+                                     l_lWritable,
+                                     NULL,
+                                     0,
+                                     PA_SEEK_RELATIVE );
+
+            l_ptrOperation = pa_stream_update_timing_info( l_ptrStream->outStream,
+                                                           NULL,
+                                                           NULL );
+            pa_threaded_mainloop_unlock( l_ptrStream->mainloop );
+
+            while( pa_operation_get_state( l_ptrOperation ) == PA_OPERATION_RUNNING)
+            {
+                usleep(100);
```

Comment:
Do we need to do error handling in this loop as well, like in:
https://github.com/pulseaudio/pulseaudio/blob/7f4d7fcf5f6407913e50604c6195d0d5356195b1/src/pulse/simple.c#L399

Reply:
I made gigantic macro PA_PULSEAUDIO_IS_ERROR which should be used everywhere but now it's on blocking API